(Feature) Report Payouts - New condition selection

gabitoesmiapodo opened the pull request:
Closes #511 (when this and #571 are finished).
Closes #514

As with #571, this is not finished and it's missing a lot of logic, which @mariano-aguero will add later, so I'm making this a draft.

- There are a lot of unused vars.
- Condition selection "works", but it's pretty buggy so it'll need fixing.
- Filters don't work.
- Search should work as it worked before.

@mariano-aguero take a look at https://zpl.io/beQ874P to know what's expected from this.

**Note:** I think https://github.com/gnosis/conditional-tokens-explorer/issues/544 has the same requirements than we have in the design's notes, so close it if that's the case.

![Nov-03-2020 13-43-49](https://user-images.githubusercontent.com/4015436/98014716-c2033c00-1dda-11eb-8830-82427c433b04.gif)

mariano-aguero commented:
I think I'm done here , @hpmaxi your call now

gnosis-info commented:
Travis automatic deployment:
 https://pr582--conditionaltokensexplorer.review.gnosisdev.com

elena-zh commented:
I went through the changes. And they LGTM. However, here are some issues to fix/discuss: 
1. Is it okay that a not connected user sees the 'already' applied filters to the form (and has no possibility to 'play' with the form until he/she open filters section)? 
![not connected.jpg](https://images.zenhubusercontent.com/5f55ec970a2e890ec10f7778/9cfb4c0a-7775-41d5-94e7-3c51725e7517)
2. 'Payout' field looks a bit different than on the mock-up. In addition, the value in the 'new' field is left-aligned
![payout.jpg](https://images.zenhubusercontent.com/5f55ec970a2e890ec10f7778/606a0b4d-a9d8-47d4-9eac-cca205b2ae43)
3. When a condition is  successfully resolved, the conditions section is not auto-updated and the already resolved condition is still displayed there
4. Condition ID is not auto-preselected when navigate to the Report Payout page from the Conditions list/a condition detail
5. In addition to the p.4: it would be nice to disable 'Report Payout' option in the conditions list for those conditions which can't be resolved by a user
6. When I select a condition with the help of filters, and then switch the filters' toggle off, all the changes I have done on the form are reset (condition selection, payouts' fields values). It would be nice whether to throw 'Are you sure...' pop-up here or save the selected condition/values in the fields when hiding filters
7. This UI issue is related to the #571 as well: search option is not fully colored (some missing pixels at the top and from the right border)
![is not fully coloured.jpg](https://images.zenhubusercontent.com/5f55ec970a2e890ec10f7778/970cb7ee-758d-42c5-949d-21c6db8c9b33)

mariano-aguero commented:
> I went through the changes. And they LGTM. However, here are some issues to fix/discuss:
> 
> 1. Is it okay that a not connected user sees the 'already' applied filters to the form (and has no possibility to 'play' with the form until he/she open filters section)?
>    ![not connected.jpg](https://camo.githubusercontent.com/da5f2484c372bd24ca88a267348a7854320f71796254d4eeed27eb00d91cca3c/68747470733a2f2f696d616765732e7a656e68756275736572636f6e74656e742e636f6d2f3566353565633937306132653839306563313066373737382f39636662346330612d373737352d343164352d393465372d336335313732356537353137)


For me it isn't so relevant, in order to report a condition the user's address has to be the oracle and the condition must be in an open state

> 2. 'Payout' field looks a bit different than on the mock-up. In addition, the value in the 'new' field is left-aligned
>    ![payout.jpg](https://camo.githubusercontent.com/8a732a76014b0bed2b106db595e7068b9ba302668c03b8efe519187c8a24f580/68747470733a2f2f696d616765732e7a656e68756275736572636f6e74656e742e636f6d2f3566353565633937306132653839306563313066373737382f36303661306234642d613964382d343764342d396561632d636361323035623261653433)

Gabriel is already aware of that and will make the necessary modifications

> 3. When a condition is  successfully resolved, the conditions section is not auto-updated and the already resolved condition is still displayed there

Mmmm exist a `refetch` method, I will check it out

> 4. Condition ID is not auto-preselected when navigate to the Report Payout page from the Conditions list/a condition detail

I will take a look

> 5. In addition to the p.4: it would be nice to disable 'Report Payout' option in the conditions list for those conditions which can't be resolved by a user

I will take a look

> 6. When I select a condition with the help of filters, and then switch the filters' toggle off, all the changes I have done on the form are reset (condition selection, payouts' fields values). It would be nice whether to throw 'Are you sure...' pop-up here or save the selected condition/values in the fields when hiding filters

I will take a look

> 7. This UI issue is related to the #571 as well: search option is not fully colored (some missing pixels at the top and from the right border)
>    ![is not fully coloured.jpg](https://camo.githubusercontent.com/f391a55798397028369dfeea3d03d458f4ff521c9ac3b4dc5dba4959257baa52/68747470733a2f2f696d616765732e7a656e68756275736572636f6e74656e742e636f6d2f3566353565633937306132653839306563313066373737382f39373063623765652d373538642d343263352d393439642d323163366462386339623333)

Maybe you can move this comment to the #571

elena-zh commented:
@mariano-aguero , thanks for clarifications! I have added the 7th issue to the #571

gnosis-info commented:
Travis automatic deployment:
 https://pr582--conditionaltokensexplorer.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr582--conditionaltokensexplorer.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr582--conditionaltokensexplorer.review.gnosisdev.com

mariano-aguero commented:
> 3. When a condition is  successfully resolved, the conditions section is not auto-updated and the already resolved condition is still displayed there

Mmm this is a huge problem, because the list is already been updated, but the subgraph is not indexing faster enough. The solution requires for every element to check if the condition triggered and event  `ConditionResolution`, and remove that item from the list, is not so easy and could and may incur in performance issues. I think we can trust that the form will not allow to resolve an already resolved condition. Also you are allowed to filtering by conditions already resolved, so ...

> 4. Condition ID is not auto-preselected when navigate to the Report Payout page from the Conditions list/a condition detail

Done

> 5. In addition to the p.4: it would be nice to disable 'Report Payout' option in the conditions list for those conditions which can't be resolved by a user

Done, also in the condition details

> 6. When I select a condition with the help of filters, and then switch the filters' toggle off, all the changes I have done on the form are reset (condition selection, payouts' fields values). It would be nice whether to throw 'Are you sure...' pop-up here or save the selected condition/values in the fields when hiding filters

Done

hpmaxi commented on src/pages/ReportPayouts/Contents.tsx:
```diff
@@ -93,54 +66,129 @@ export const Contents: React.FC = () => {
+      <SelectableConditionTable
+        allowToDisplayOnlyConditionsToReport={true}
+        onClearSelection={() => {
+          if (payouts.length > 0 || conditionId) {
+            if (
+              window.confirm(
+                `Are you sure you want to clear the filters? The changes you made will be lost?`
+              )
+            ) {
+              setPayouts([])
+              setConditionId('')
```
This could also use `isPayoutsEmpty`?

mariano-aguero replied:
Code removed

hpmaxi reviewed (approved):
Looks great!. Good job and use of Array.prototype methods!

gnosis-info commented:
Travis automatic deployment:
 https://pr582--conditionaltokensexplorer.review.gnosisdev.com

gabitoesmiapodo commented:
@mariano-aguero @elena-zh 

2. -> Should be fixed.
7. -> The only way I could reproduce that is by zooming in (CTRL+), but I couldn't see the issue otherwise. Anyways, I added a green background to the wrapping component, that probably will fix the issue.

BTW, I removed the filter removal. We assume that "switching off" filters will clear ("turn off") them, and hope that it will be obvious enough for the users. Let's see if there's some feedback about that and we can improve it later if it's necessary. I'm OK with having some safeguards, but it can be annoying if there are too many and losing some filter configuration is not as big an issue as losing a completed form.

gnosis-info commented:
Travis automatic deployment:
 https://pr582--conditionaltokensexplorer.review.gnosisdev.com

elena-zh commented:
@mariano-aguero , @gabitoesmiapodo , thanks you for you job and clarifications.  Here are my notes: 
2. The Payout fields LGTM now, however, there are issues found: 
2.1 It is allowed to enter a negative amount into the field (on the current version '-' is not allowed to be entered)
![minus](https://user-images.githubusercontent.com/70885163/98807195-5862dd80-242b-11eb-800c-68c951908ca9.jpg)
![negative result](https://user-images.githubusercontent.com/70885163/98807198-59940a80-242b-11eb-819c-78c849a13fcc.jpg)

2.2. there is no 0.00 placeholder in the fields when they are empty (they are exist in the current version)
2.3. there is no validation 'At least one payout must be positive'  when clear out all the fields (like it is on the current version)
![at least](https://user-images.githubusercontent.com/70885163/98807394-a24bc380-242b-11eb-8ad5-4d61f1300a1d.jpg)

3.  Thank you for the investigation. Let leave it as it is for now. BTW, I have created a separate low-priority enhancement for this issue in case if it comes from user's feedback. 

4. The issue is fixed. However, the screen is blinking on the Report payout page right after navigating to the page. See the [video ](https://drive.google.com/file/d/1SJB_9v_nKw8HR82dFBdHs1ZkQQ7HiXf4/view)

5. Great!

6.  I have added a comment [here (see p.8)](https://github.com/gnosis/conditional-tokens-explorer/pull/598#issuecomment-725344723)

7. The issue is reproducible on a certain screen resolution. I have added a comment [here ](https://github.com/gnosis/conditional-tokens-explorer/issues/600#issuecomment-725368200)

elena-zh commented:
@gabitoesmiapodo , @mariano-aguero is this PR merged to Dev?
I cannot see any changes there
![image.png](https://images.zenhubusercontent.com/5f55ec970a2e890ec10f7778/a8e3b0df-b9a7-4f5f-9ca5-1edcb7c985e3)

gabitoesmiapodo commented:
@elena-zh It's merged into `/new-design-integration`, which should be updated with the latest changes from `/develop`

elena-zh commented:
Verified and LGTM. Some low priority issues:
1. Conditions section is not auto-updated when a condition is resolved #601 
2. not to show 'At least one payout must be positive' warning until clear out all the payouts fields #615 
3. Disable sorting by Reporter/Oracle and QuestionID in the select condition section #647 
4. Report button should be enabled and call connection pop-up for a not logged in user #652 
5. Sometimes empty reported array is displayed after successful report payouts #642
